modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/AddWriteLinkingWiInvokeClosure.java

```diff
@@ -0,0 +1,158 @@
+ *
+ * <p>See {@link AbstractPageMemoryMvPartitionStorage} about synchronization.
+ *
+ * <p>Operation may throw {@link StorageException} which will cause form {@link BplusTree#invoke(Object, Object, InvokeClosure)}.
+ */
+class AddWriteLinkingWiInvokeClosure extends AddWriteInvokeClosure {
+    private final PersistentPageMemoryMvPartitionStorage persistentStorage;
+
+    private final FreeList freeList;
+
+    private long wiListHeadLink;
+    private long newWiListHeadLink;
+
+    AddWriteLinkingWiInvokeClosure(
+            RowId rowId,
+            @Nullable BinaryRow row,
+            UUID txId,
+            int commitZoneId,
+            int commitPartitionId,
+            PersistentPageMemoryMvPartitionStorage storage
+    ) {
+        super(rowId, row, txId, commitZoneId, commitPartitionId, storage);
+
+        persistentStorage = storage;
+
+        freeList = storage.renewableState.freeList();
+    }
+
+    @Override
+    public void call(@Nullable VersionChain oldRow) throws IgniteInternalCheckedException {
+        wiListHeadLink = persistentStorage.lockWriteIntentListHead();
+        newWiListHeadLink = wiListHeadLink;
+
+        try {
+            super.call(oldRow);
+        } finally {
+            persistentStorage.updateWriteIntentListHeadAndUnlock(newWiListHeadLink);
+        }
+    }
+
+    @Override
+    protected RowVersion insertFirstRowVersion() {
+        assert persistentStorage.writeIntentHeadIsLockedByCurrentThread();
+        assert persistentStorage.writeIntentListHead() == wiListHeadLink
+                : "Expected WI list head link " + wiListHeadLink + " but was " + persistentStorage.writeIntentListHead();
+
+        WiLinkableRowVersion newVersion = insertRowVersion(NULL_LINK, wiListHeadLink, NULL_LINK);
+
+        newWiListHeadLink = newVersion.link();
+
+        updateWiListLinks(newVersion);
+
+        return newVersion;
+    }
+
+    @Override
+    protected RowVersion insertAnotherRowVersion(VersionChain oldRow, @Nullable RowVersion existingWriteIntent) {
+        assert persistentStorage.writeIntentHeadIsLockedByCurrentThread();
+        assert persistentStorage.writeIntentListHead() == wiListHeadLink
+                : "Expected WI list head link " + wiListHeadLink + " but was " + persistentStorage.writeIntentListHead();
+
+        boolean replacingExistingWriteIntent = oldRow.isUncommitted();
+        assert replacingExistingWriteIntent == (existingWriteIntent != null);
+
+        long newNextWiLink;
+        long newPrevWiLink;
+        if (replacingExistingWriteIntent) {
+            newNextWiLink = existingWriteIntent.operations().nextWriteIntentLink(wiListHeadLink);
+            newPrevWiLink = existingWriteIntent.operations().prevWriteIntentLink();
+        } else {
+            newNextWiLink = wiListHeadLink;
+            newPrevWiLink = NULL_LINK;
+        }
+
+        WiLinkableRowVersion newVersion = insertRowVersion(oldRow.newestCommittedLink(), newNextWiLink, newPrevWiLink);
```

Comment:
What does this method do? If it actually inserts an entire row's payload into data pages then I would not recommend to do it while holding an exclusive lock. This will lead to potential contention. Operations under lock should be as trivial as possible.

What if we insert the row with `NULL/NULL` first, and only then acquire lock and update its links? How many changes would that introduce to the code?